content/blog/typestate-pattern/index.md

```diff
@@ -0,0 +1,303 @@
+   the code easier to understand and maintain.
+2. **Open/Closed Principle**: Adding a new state or behavior is simple.
+   Implementing a new state does not require modifying existing ones.
+3. **Strong Type Guarantees**: Illegal states or transitions are managed at
+   compile-time. The Rust compiler will produce an error if a required behavior
+   for a state is not implemented.
+
+However, this solution uses dynamic dispatch (`Box<dyn DoorState>`),
+which means that the method calls are resolved at runtime.
+In embedded systems or performance-critical applications, dynamic dispatch can
+be forbidden due to its runtime overhead.
+
+Here's a version with static dispatch:
+
+```rust
+trait DoorState {
+    fn open(self) -> Self;
+    fn lock(self) -> Self;
+}
+
+struct Locked;
+struct Unlocked;
+struct ChildLocked;
+
+impl DoorState for Locked {
+    fn open(self) -> Self {
+        println!("Door is locked, can't open");
+        Locked
+    }
+
+    fn lock(self) -> Self {
+        println!("Door is already locked");
+        Locked
+    }
+}
+
+impl DoorState for Unlocked {
+    fn open(self) -> Self {
+        println!("Opening door");
+        Unlocked
+    }
+
+    fn lock(self) -> Self {
+        println!("Locking door");
+        Locked
+    }
+}
+
+impl DoorState for ChildLocked {
+    fn open(self) -> Self {
+        println!("Opening door from the outside only");
+        ChildLocked
+    }
+
+    fn lock(self) -> Self {
+        println!("Locking door");
+        Locked
+    }
+}
+
+// The Door struct is now generic over the state.
+// This allows the state to be statically dispatched.
+struct Door<S: DoorState> {
+    state: S,
+}
+
+impl Door<Unlocked> {
+    fn open(self) -> Door<Unlocked> {
+        Door {
+            state: self.state.open(),
+        }
+    }
+
+    fn lock(self) -> Door<Locked> {
+        Door {
+            state: self.state.lock(),
+        }
+    }
+}
+
+impl Door<Locked> {
+    fn open(self) -> Door<Locked> {
+        Door {
+            state: self.state.open(),
+        }
+    }
+
+    fn lock(self) -> Door<Locked> {
+        Door {
+            state: self.state.lock(),
+        }
+    }
+}
+
+impl Door<ChildLocked> {
+    fn open(self) -> Door<ChildLocked> {
+        Door {
+            state: self.state.open(),
+        }
+    }
+
+    fn lock(self) -> Door<Locked> {
+        Door {
+            state: self.state.lock(),
+        }
+    }
+}
+```
+
+([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8b5b0b0b0b0b0b0b0b0b0b0b0b0b0b))
+
+
+Here's how to use it:
+
+```rust
+let door = Door { state: Unlocked };
+let door = door.lock();
+// This will not compile
+// Error: method not found in `Locked`
+// A locked door cannot be opened by pulling the handle
+let door = door.open(); 
+```
+
+And that's the typestate pattern in Rust!
+If it compiles, it works. 
+
+## Conclusion
+
+As it turns out, Rust is not just good for systems programming but also for
```

Comment:
```suggestion
As it turns out, Rust is not only good for systems programming but also for
```

Reply:
Removed that sentence because I felt it was a bit cheesy. 😅